### PR TITLE
Move handling of `root` option from serializer to adapter.

### PR DESCRIPTION
**Pull request by beauby**
This option is clearly an adapter concern and it is only used in the Json adapter.
+ Custom `root` are not subject to inflection anymore.

Most tests are failing because they are directly testing the root value on the serializers. Will fix.

### CODE REVIEW

File: test/adapter/json_api/belongs_to_test.rb [hunk 1]
```diff
@@ -38,7 +38,11 @@ def test_includes_post_id
           end
 
           def test_includes_linked_post
-            @adapter = ActiveModel::Serializer::Adapter::JsonApi.new(@serializer, include: [:post])
+            actual = ActiveModel::SerializableResource.new(
```
Comment (bf4) [inline test/adapter/json_api/belongs_to_test.rb]: use the `serializable` helper to for less words, if you want

File: lib/active_model/serializer/array_serializer.rb [hunk 2]
```diff
@@ -25,11 +24,6 @@ def initialize(resources, options = {})
         @meta_key = options[:meta_key]
       end
 
-      def json_key
-        key = root || serializers.first.try(:json_key) || object.try(:name).try(:underscore)
-        key.try(:pluralize)
-      end
```
Comment (bf4) [inline lib/active_model/serializer/array_serializer.rb]: is this not used anywhere?

Comment (beauby) [inline lib/active_model/serializer/array_serializer.rb]: Not anymore: this PR makes it an adapter concern.

**Comment by bf4** [comment]
@beauby Would you clarify what `root` and `json_key` mean in context? I thought it was the root for each resource, which would mean it's per serialized resource, no?  (But now that we have `Serializer::type` maybe that replaces this?)

File: test/adapter/json_api/belongs_to_test.rb (hunk 1 above)
Comment (beauby) [inline test/adapter/json_api/belongs_to_test.rb]: Is the helper globally available? I think it should be. If so, I'll update these tests.

**Comment by beauby** [comment]
@bf4: No, they were only referring to the root of the document, that's why I believe it's an adapter concern. For the "roots" of resources, we actually use the association's key.

**Comment by beauby** [comment]
Updated.

File: test/action_controller/serialization_test.rb [hunk 3]
```diff
@@ -171,7 +171,7 @@ def test_render_array_using_custom_root
         with_adapter :json do
           get :render_array_using_custom_root
         end
-        expected = { custom_roots: [{ name: 'Name 1', description: 'Description 1' }] }
+        expected = { custom_root: [{ name: 'Name 1', description: 'Description 1' }] }
```
Comment (bf4) [inline test/action_controller/serialization_test.rb]: this raises a huge red flag for me.  Why is this changing?

Comment (beauby) [inline test/action_controller/serialization_test.rb]: I believe it should not have been there in the first place. Modifying explicitly supplied options is such an antipattern. I could move this change to an other PR though.

File: lib/active_model/serializer/adapter/json.rb [hunk 4]
```diff
@@ -15,6 +15,24 @@ def serializable_hash(options = nil)
...
+                  serializer.object.try(:name)
+                end
+              key.to_s.pluralize
+            else
+              serializer.object.class.model_name.to_s
+            end
+          end
+
+          root.to_s.underscore.to_sym
+        end
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json.rb]: vs. 

```diff
   def json_key
-      root || object.class.model_name.to_s.underscore
-    end
```

and

```diff
-      def json_key
-        key = root || serializers.first.try(:json_key) || object.try(:name).try(:underscore)
-        key.try(:pluralize)
-      end
```

?

Shouldn't this just be using the Serializer `type` or am I tired?
I'm not convinced

Comment (beauby) [inline lib/active_model/serializer/adapter/json.rb]: Yes. Highly possible that the serializer's `type` is the right thing to use here. Main point of this PR was that an option that is specific to the `Json` adapter should not clutter the `Serializer` and `Adapter::Base` classes with lots of overcomplicated code.

Comment (beauby) [inline lib/active_model/serializer/adapter/json.rb]: Note that `type` is nil unless explicitly specified though.

**Comment by NullVoxPopuli** [comment]
is this still applicable?

**Comment by bf4** [comment]
Done, iirc

B mobile phone

**Comment by remear** [comment]
Can this be closed then?

**Comment by beauby** [comment]
Seems not: https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializable_resource.rb#L5
As you can see, `:root` is not an adapter option yet.

**Comment by bf4** [comment]
@beauby I would love if you'd come back to this one. Shouldn't be to hard to finish.

**Comment by beauby** [comment]
Will take care of it over the weekend!

**Comment by beauby** [comment]
Closing in favor of https://github.com/rails-api/active_model_serializers/pull/1782.

**Pull request closed without merging**